src/Storages/Distributed/DistributedAsyncInsertDirectoryQueue.cpp

```diff
@@ -602,9 +629,20 @@ void DistributedAsyncInsertDirectoryQueue::processFilesWithBatching(const Settin
             }
             catch (const Exception & e)
             {
-                if (isDistributedSendBroken(e.code(), e.isRemoteException()))
+                incrementFilesRetryNum(file_path);
+
+                if (isDistributedSendBroken(e.code(), e.isRemoteException()) || (max_retries != 0 && files_retry[file_path] == max_retries))
                 {
-                    markAsBroken(file_path);
+                    if (const auto it = files_retry.find(file_path); it != files_retry.end())
+                    {
+                        files_retry.erase(file_path);
+                        /// file might grow without a limit, need delete.
+                        markAsSend(file_path);
+                        LOG_TRACE(log, "Finished retry {}, success delete {}", max_retries, file_path);
+                    } else
+                    {
+                        markAsBroken(file_path);
+                    }
```

Comment:
I have multiple problems with this:
1. We don't actually send the batch here, so I think this is not a good logic here.
2. When will we retry `file_path`? We pop its value from the queue, thus we will never retry. But you can try to verify with a test.

Reply:
<img width="890" alt="image" src="https://github.com/user-attachments/assets/7de6ff98-68e1-40fa-a6e7-c5600b45f848">
while (!pending_files.isFinished()) this will keep retrying files that are not marked broken

Reply:
I think we pop files from `pending_files` and we don't push them back unless we throw. I might be wrong, but on line 586 we always pop a new path `pending_files.tryPop(file_path)`, and we only push them back on line 684. So this means if we just do `continue` here, then that files is lost. If I am wrong, then please correct me, but not with a pasted image please.

Reply:
I am studying your ideas and making some modifications. Thank you for your suggestions

Reply:
after the exception is throw on line 651, the file is pushed, and the file is not lost

Reply:
Where do we throw exception on line 651? I cannot see that.